apis/python/src/tiledbsoma/io/shaping.py

```diff
@@ -361,83 +337,198 @@ def _recurse(
+    ) -> dict[str, Any]:
         """Applies ``_treewalk`` to the requested child element.
 
         Args:
             parent: The parent item to get child from.
             name: Name of the child element to attempt to visit.
-            sizing_args: Arguments to pass to the visitor.
+            kwargs: as in ``_treewalk``.
 
         Returns:
-            The status of the call. If the child exists, return the status of
-            ``_treewalk`, otherwise return ``True``.
+            A nested map with results from all leaf nodes.
         """
-        if name in parent:
+        if node_name is not None and node_name in parent:
             return _treewalk(
-                parent[name].uri,
-                node_name=name,
-                visitor=visitor,
-                args=sizing_args,
-                context=context,
+                uri=parent[node_name].uri,
+                node_name=node_name,
+                leaf_visitor=leaf_visitor,
+                **kwargs,
             )
-        return True
+        return {}
 
-    with tiledbsoma.open(uri, context=context) as item:
+    # Don't abort the treewalk, except in case of exceptions of course.
+    # When users want to see if an experiment is upgradeable, they need
+    # to be shown *all* the component arrays which have issues, not just
+    # the first such.
+    with tiledbsoma.open(uri, context=kwargs["context"]) as item:
 
+        # Non-terminal
         if isinstance(item, tiledbsoma.Experiment):
-            status = _recurse(item, "obs", args)
-            status &= _recurse(item, "ms", args)
-            return status
+            retval = {}
+            retval["obs"] = _recurse(item, node_name="obs", **kwargs)
+            retval["ms"] = _recurse(item, node_name="ms", **kwargs)
+            return retval
 
+        # Non-terminal
         if isinstance(item, tiledbsoma.Measurement):
-            status = _recurse(item, "var", args)
-            for coll_name in ["X", "obsm", "obsp", "varm", "varp"]:
-                args["coll_name"] = coll_name
-                status &= _recurse(item, coll_name, args)
-            return status
+            retval = {}
+            retval["var"] = _recurse(item, node_name="var", **kwargs)
+            for iter_coll_name in ["X", "obsm", "obsp", "varm", "varp"]:
+                kwargs["coll_name"] = iter_coll_name
+                retval[iter_coll_name] = _recurse(
+                    item, node_name=iter_coll_name, **kwargs
+                )
+            return retval
 
+        # Non-terminal
         if isinstance(item, tiledbsoma.Collection):
-            status = True
+            retval = {}
             for key in item:
                 if node_name == "ms":
-                    args["ms_name"] = key
-                status &= _recurse(item, key, args)
-            return status
+                    kwargs["ms_name"] = key
+                retval[key] = _recurse(item, node_name=key, **kwargs)
```

Comment:
```suggestion
                    retval[key] = _recurse(item, node_name=key, ms_name=key, **kwargs)
                else: 
                    retval[key] = _recurse(item, node_name=key, **kwargs)
```
Optional: ymmv

Reply:
As above

Reply:
My bad. I didn't take into account that `_recurse` ... recurses